### ISSUE TITLE
syntax highligting black and white

### ISSUE THREAD

**Issue by cacplate** [neovim/neovim#2019]
Since the merge of #1820 (I "think" it is from this one since other PR were merged also), my colorscheme is loaded and the bold on certain words is present but I got no color whatsoever, only white on gray. The colorscheme is recognized and the good filetype is set. Any idea?

**Comment by fmoralesc** [comment]
What terminal and colorscheme are you using?

**Comment by cacplate** [comment]
@fmoralesc sakura, molokai

**Comment by cacplate** [comment]
It's worth noting that I tried some other themes without success except with the guardian(http://www.vim.org/scripts/script.php?script_id=1240) which is all fucked up but with colors

**Comment by cacplate** [comment]
Also, the visual mode highlighting is not present but the mode works as usual.

**Comment by fmoralesc** [comment]
Do you get this behavior in other terminals too?

**Comment by fmoralesc** [comment]
I tested sakura with nvim from HEAD, and it seems to work fine. What is your value for $TERM (mine was `xterm-256color`)? Also, it might be helpful if you provided your vimrc file.

**Comment by Pyrohh** [comment]
Sorry to interject, but does this happen since #2007 was merged?

**Comment by cacplate** [comment]
@Pyrohh you're right my bad.
@fmoralesc I tried with gnome-terminal too but nothing changed.
my $TERM is set to "xterm" and I never touched it.
```
set encoding=utf-8
set fileencoding=utf-8
set ignorecase
set undofile
set textwidth=118
set number
set hlsearch
set incsearch
set autochdir
set splitright
set expandtab
set tabstop=4
set shiftwidth=4
set softtabstop=4
set scrolloff=10
set mouse=a
set t_Co=256
set bs=indent,eol,start
set clipboard+=unnamedplus

filetype off
set rtp+=~/.nvim/bundle/Vundle.vim
call vundle#begin()
Plugin 'gmarik/vundle'
Plugin 'scrooloose/nerdtree'
Plugin 'townk/vim-autoclose'
Plugin 'tpope/vim-surround'
Plugin 'tpope/vim-repeat'
Plugin 'luochen1990/rainbow'
Plugin 'travitch/hasksyn'
Plugin 'vim-scripts/vim-auto-save'
Plugin 'vim-scripts/toggle_comment'
Plugin 'vim-scripts/sudo.vim'
Plugin 'vim-scripts/mru.vim'
Plugin 'vim-scripts/tex_mini.vim'
Plugin 'ervandew/supertab'
call vundle#end()
filetype plugin indent on

syntax on
colorscheme molokai
let g:rainbow_active = 1
let g:auto_save = 1
highlight ColorColumn ctermbg=magenta
call matchadd('ColorColumn', '\%120v', 100)
set listchars=tab:>~,nbsp:_,trail:.
set list

nnoremap ; :
nnoremap q :bdelete<CR>
nnoremap e :e<Space>
nnoremap <Space> :noh<CR>
nnoremap \ :MRU<CR>
nnoremap w :w<Space>
nnoremap n :NERDTreeToggle<CR>
nnoremap <C-\> :vsplit new<CR>
nnoremap <tab> :bn<CR>
nnoremap <S-tab> :buffers<CR>:buffer<Space>

autocmd FileType python nnoremap <CR> :!python %<CR>
autocmd FileType haskell nnoremap <CR> :!runghc %<CR>

autocmd FileType j nnoremap <CR> :!chmod +x % && ./%<CR>
autocmd FileType j :AutoCloseOff

autocmd FileType tex nnoremap <CR> :!pdflatex % && rm %:r.aux %:r.log && evince %:r.pdf<CR>
autocmd FileType tex nnoremap f :!detex % \| wc -w<CR>

autocmd FileType c nnoremap <CR> :!clang -Wall -std=c99 -lm -o %:r % && ./%:r<CR>
autocmd FileType c inoremap {{ <Space>{<CR>}<C-o>==<C-o>O
autocmd FileType c nnoremap f :g/^[^#].*[^{};]$/norm A;<CR>
```

**Comment by kurkale6ka** [comment]
This might be related to my issue #2015

**Comment by crepererum** [comment]
I have the same problem with [tomorrow](https://github.com/chriskempson/tomorrow-theme). neobundle loads it correctly but when setting the theme via `colorscheme`, I don not get any colors (only thick-thin). In contrast this works with the bundled default themes (e.g. morning).

**Comment by cacplate** [comment]
@crepererum Morning gives me colors but I'm not sure if they the good ones

**Comment by crepererum** [comment]
@cacplate Yes, as I (maybe not clearly ;) ) said, morning works. I think `has("gui_running")` (used by some themes for detection work) isn't in master anymore and therefor breaks some themes. The question now is if this change was intended or not.

**Comment by somebody32** [comment]
Same for me, don't have any colours at all with non-bundled themes

**Comment by tarruda** [comment]
For everyone experiencing this issue, what happens when you start nvim like this: `TERM=xterm-256color nvim`?

**Comment by kurkale6ka** [comment]
Hmm, `TERM=xterm-256color nvim` fixes my #2015 issue

**Comment by crepererum** [comment]
@tarruda For me there is no difference difference.

**Comment by crepererum** [comment]
It seems that some colors work while others clearly doesn't. Maybe that screenshot helps (using Tomorrow):
![screenshot from 2015-02-20 11 23 32](https://cloud.githubusercontent.com/assets/1529400/6284670/faee3374-b8f2-11e4-9c3e-12eee46d7c6c.png)

**Comment by cacplate** [comment]
XTERM=xterm-256color fixes my issue (I only tested with molokai though)
@crepererum what is your color scheme? I could test to see if it's you or the colorscheme.

**Comment by tarruda** [comment]
I suppose I can set some 256 color capabilities for `TERM=xterm`. It seems many terminals don't set xterm-256color by default even though they support it.

**Comment by tarruda** [comment]
@crepererum does this patch fixes the problem?

```diff
diff --git a/src/nvim/tui/tui.c b/src/nvim/tui/tui.c
index 5951408..7c4d157 100644
--- a/src/nvim/tui/tui.c
+++ b/src/nvim/tui/tui.c
@@ -717,6 +717,19 @@ static void fix_terminfo(TUIData *data)
         "\x1b[?2004l");
   }

+#define XTERM_SETAF \
+  "\x1b[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m"
+#define XTERM_SETAB \
+  "\x1b[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m"
+
+  if (!strcmp(term, "xterm")) {
+    // probably every modern terminal that sets TERM=xterm supports 256
+    // colors(eg: gnome-terminal).
+    unibi_set_num(ut, unibi_max_colors, 256);
+    unibi_set_str(ut, unibi_set_a_foreground, XTERM_SETAF);
+    unibi_set_str(ut, unibi_set_a_background, XTERM_SETAB);
+  }
+
   if (os_getenv("NVIM_TUI_ENABLE_CURSOR_SHAPE") == NULL) {
     goto end;
   }
@@ -748,10 +761,8 @@ end:
       "\x1b[?1002l\x1b[?1006l");
   unibi_set_if_empty(ut, unibi_cursor_address, "\x1b[%i%p1%d;%p2%dH");
   unibi_set_if_empty(ut, unibi_exit_attribute_mode, "\x1b[0;10m");
-  unibi_set_if_empty(ut, unibi_set_a_foreground,
-      "\x1b[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m");
-  unibi_set_if_empty(ut, unibi_set_a_background,
-      "\x1b[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m");
+  unibi_set_if_empty(ut, unibi_set_a_foreground, XTERM_SETAF);
+  unibi_set_if_empty(ut, unibi_set_a_background, XTERM_SETAB);
   unibi_set_if_empty(ut, unibi_enter_bold_mode, "\x1b[1m");
   unibi_set_if_empty(ut, unibi_enter_underline_mode, "\x1b[4m");
   unibi_set_if_empty(ut, unibi_enter_reverse_mode, "\x1b[7m");
```

**Comment by jszakmeister** [comment]
> It seems many terminals don't set xterm-256color by default even though they support it.

That's definitely true.  I have this hack in my vimfiles for gnome-terminal:

```vim
" For some reason, gnome-terminal says xterm-color even though it supports
" xterm-256color.
if !has("gui_running") && $COLORTERM == "gnome-terminal" && &t_Co <= 16
    if $TERM == "xterm"
        " Override the default xterm setting of Gnome Terminal so that
        " Powerline works.
        let $TERM = "xterm-256color"
    endif
    set t_Co=256
endif
```

**Comment by tarruda** [comment]
> That's definitely true. I have this hack in my vimfiles for gnome-terminal:

Hmm, I didn't knew about the $COLORTERM environment variable. I will use it instead of simply cheking $TERM :)

**Comment by tarruda** [comment]
Included the patch to #1997

**Comment by crepererum** [comment]
@tarruda OK, that was clearly my fault. I set $XTERM instead of $TERM. So when setting $TERM=xterm-256color, it works. Thing is that gnome-terminal does not set this. The breaking change is that vim works with `set t_Co=256` while neovim now (=master) ignores this kind of hack.

**Comment by fmoralesc** [comment]
I also had that issue with gnome-terminal in vim , I can confirm it won't
use xterm-256color by default. Detecting it with COLORTERM is safe.

**Comment by tarruda** [comment]
I merged #1997 which includes the $COLORTERM check. Can you verify if this fixes the issue?

**Comment by crepererum** [comment]
Now I get the following results:
 - `TERM=xterm` => no colors (correctly detected by using it)
 - `TERM=foo` (random string) or `TERM=` (empty) => colors (autodetection?)
 - `TERM=xterm-256color` => colors (correctly detected by using it)
 - `TERM unset` => SIGSEGV (this should not happen)

Btw: My gnome-terminal does not set `COLORTERM` (gnome 3.14).

**Comment by mseeber** [comment]
On debian testing, setting COLORTERM solves the issue outside of tmux.

I should add, that $COLORTERM doesn't always seem to be available in gnome-terminal on debian testing, which has $TERM set to "xterm".

This also causes issues with tmux and the less intrusive solution I found is to use a bash alias setting TERM for tmux and nvim, as setting colorterm

    alias tmux='TERM=xterm-256color /usr/bin/tmux -2'
    alias nvim='TERM=xterm-256color /path/to/neovim/bin/nvim'

Adding a bash alias for nvim with the $COLORTERM variable works in this case outside of tmux, but not in tmux.

I think tough, this is not an issue with nvim, more of a problem with the terminal.

**Comment by tarruda** [comment]
@mseeber [Here's](https://github.com/neovim/neovim/blob/6293b5513c54557bbf56077374df83cb52344fa3/src/nvim/tui/tui.c#L725) the code that checks for `$COLORTERM`, meaning it will only override terminfo settings if $TERM is xterm and $COLORTERM is set.

I think the problem is that tmux sets TERM to "screen" by default, but if you configure to use screen-256color it might fix the issue. Try adding this line to tmux.conf: `set -g default-terminal "screen-256color"`

**Comment by cacplate** [comment]
`$COLORTERM` works for me but only to replace `TERM=xterm-256color`. It doesn't work with my shortcut since the execute command doesn't load my zshrc. I can confirm that the `set t_Co=256` doesn't work anymore for consistency of 256 colors regardless of the way to open neovim

**Comment by mseeber** [comment]
> meaning it will only override terminfo settings if $TERM is xterm and $COLORTERM is set.

what happens on systems without support for 256 colors that set $TERM=xterm and $COLORTERM to some other value, like 16 colors? (is there even such a case?)

> I think the problem is that tmux sets TERM to "screen" by default

yes, but i still have to set $TERM for tmux, since tmux seems to parse $TERM and adjusts it's colors, regardles according to $TERM.

so i end up with this in my .bsah_aliases:
    alias tmux='TERM=xterm-256color /usr/bin/tmux'
    export COLORTERM=xterm-256color

.tmux.conf
    set -g default-terminal "screen-256color"

This works in my case for vim and nvim in- and outside of tmux.

So outside of tmux:
    $TERM=xterm
    $COLORTERM=xterm-256color

inside tmux
    $TERM=screen-256color
    $COLORTERM=xterm-256color

**Comment by tarruda** [comment]
@mseeber I've included a check for TERM==screen when COLORTERM is set, so it probably covers the default usage of gnome-terminal + tmux

**Comment by mseeber** [comment]
Ah, I tested this and now the line in .tmux.conf is no longer required.

Bugreports on the debian bugtracker indicate, that there are some problems in tmux detecting the number of colors, so i guess everything is fine here.

The setup which works for me now is:

.bash_aliases:

    alias tmux='TERM=xterm-256color /usr/bin/tmux'                                                                                                 
    export COLORTERM=xterm-256color

**Comment by tarruda** [comment]
@mseeber are you saying COLORTERM isn't set automatically for you?  I'm on ubuntu 12.04 and it is set automatically for urxvt, gnome terminal and xfce terminal

**Comment by somebody32** [comment]
the same for me on iterm2/ssh to 14.10 + tmux. `export COLORTERM=xterm-256color` brought colours back (nvim -v: Commit: 3e292316846708dfc1153d55fbd5f012c2da8a35)

**Comment by mseeber** [comment]
Yes, it is not set up per default, I am on debian jessie, using gnome-terminal.

**Comment by crepererum** [comment]
Gnome terminal does NOT set $COLORTERM anymore. Ref: https://bugs.launchpad.net/byobu/+bug/1382598

**Comment by mseeber** [comment]
So this means these steps can't reliably determine in a portable way, if the terminal supports 256 colors? Any idea how to resolve or handle this from the neovim perspective?

Or is it just save to assume, that every Terminal exporting $TERM=xterm supports 256 colors?

**Comment by crepererum** [comment]
It seems that even `tput colors` relays on the correct variable. And I guess that's one of the reasons why vim has the `t_Co`variable to overwrite whatever the terminal tells you.

**Comment by cmcaine** [comment]
If you force 256 colour behaviour you're likely to break the terminals that are actually both xterm-16color and xterm-256color (at least iTerm2 and Gnome Terminal, I think).

And if you break 16color you break a lot of colorschemes that demand precise terminal colours (e.g. Solarized).

**Comment by mseeber** [comment]
Honestly, the more i read about it, the more I get the feeling, that (graphical) terminals on linux distros are totally a mess. Especially, when it comes to colours, results depend specifically on implementation specific quirks and bugs  and on custom settings. There is no silver bullet for this, as $TERM=xterm could mean anything, there is probably no other way than having the user configure his terminals correctly by hand.

**Comment by jszakmeister** [comment]
> Honestly, the more i read about it, the more I get the feeling, that (graphical) terminals on linux distros are totally a mess.

That's not a feeling.  It's fact. :-)  Paul Evans has been desperately trying to fix the situation, but there's lots of historical precedence and compatibility issues to cope with.  I'm glad he's fighting he good fight though.

> There is no silver bullet for this, as $TERM=xterm could mean anything, there is probably no other way than having the user configure his terminals correctly by hand.

Very true.  You might want to take a look at the [docs in the vim-fixkey](https://github.com/drmikehenry/vim-fixkey/blob/master/doc/fixkey.txt#L101) plugin.  It has some helpful bits on what to do for terminals to make regular vim work better with all of your keys.  It doesn't talk about colors, but you can see the kind of mess we've all been left with.

**Comment by cacplate** [comment]
Fixed, I just found #2334 that is just what I wanted.
